api/document/serializers/content.py

```diff
@@ -123,19 +127,34 @@ def get_content_type(self, instance: NestableAnnotation):
         return self.CONTENT_TYPE
 
     def get_inlines(self, instance: NestableAnnotation):
+        if self.IS_LEAF:
+            return []
         return NestedAnnotationSerializer(
             instance.children, context=self.context, many=True).data
 
     def get_text(self, instance: Annotation):
         return self.doc_node_text[instance.start:instance.end]
 
+    def to_internal_value(self, data: JsonDict) -> JsonDict:
+        inlines = [] if self.IS_LEAF else data.get('inlines', [])
+        result = super().to_internal_value(data)
+        result['content_type'] = self.CONTENT_TYPE
+        result['inlines'] = []
+        if inlines:
+            serializer = NestedAnnotationSerializer()
+            result['inlines'].extend([
+                serializer.to_internal_value(item) for item in inlines
+            ])
+        if self.IS_LEAF:
+            serializer = serializers.CharField(trim_whitespace=False)
+            result['text'] = serializer.to_internal_value(data.get('text'))
```

Comment:
This is deserializing into a different shape than what was originally serialized, which seems fishy. I wonder if we should be tracking offsets _here_.

Reply:
Wait, it is?  I was just using `serializers.CharField` to raise any validation errors if e.g. the payload provided a non-string value for the `text` key. My intent was definitely not to deserialize into a different shape than what was originally serialized, at least.

Reply:
Er, that said, I _am_ deserializing into a _subset_ of the shape that was originally serialized, e.g. non-leaf nodes don't have `text` on them, since that duplicates the content of their inlines. (That said, this goes back to the notion of documenting this endpoint, because it's unclear whether doing a `GET` and then modifying a non-leaf content node's `text` field, or even a child node's `text` field, should do anything. In this case I decided it shouldn't, since the "real" content is made up of all the text nodes, and the text of non-leaf/child nodes is really a lossy representation of what's inside. But maybe we should discuss this more if we both have different ideas of how the `PUT` endpoint should work.)